packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -238,106 +262,106 @@ contract Bridge is EssentialContract, IBridge {
+        // Refund the processing fee and fee to refund
+        unchecked {
+            // `receipt.feePaid > _message.fee` is only true if we have old data where
+            // receipt.feePaid bytes are used as an address
+            local.remainingFee = receipt.feePaid == 0 || receipt.feePaid > _message.fee
+                ? _message.fee
+                : _message.fee - receipt.feePaid;
+        }
 
-            // Process message differently based on the target address
-            if (
-                _message.to == address(0) || _message.to == address(this)
-                    || _message.to == signalService
-            ) {
-                // Handle special addresses that don't require actual invocation but
-                // mark message as DONE
-                refundAmount = _message.value;
-                _updateMessageStatus(msgHash, Status.DONE);
-            } else {
-                uint256 gasLimit;
-                if (msg.sender == _message.destOwner) {
-                    // Use the remaining gas if called by a the destOwner, else
-                    // use the specified gas limit.
-                    gasLimit = gasleft();
-                } else {
-                    // The "1/64th rule" refers to the gasleft at the time the call is made. When a
-                    // contract makes a call to another contract, it can only forward 63/64 of the
-                    // gas remaining (gasleft) at that moment, ensuring that there is always some
-                    // gas reserved for the calling contract to complete its execution after the
-                    // called contract finishes. This does not necessarily relate to the gas amount
-                    // specified in the call itself, but rather to the actual remaining gas at the
-                    // time of the call.
-                    //
-                    // See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md
-                    if (_message.gasLimit > (gasleft() * 63) >> 6) revert B_NOT_ENOUGH_GASLEFT();
-
-                    gasLimit = _message.gasLimit;
-                }
+        local.refundAmount += local.remainingFee;
 
-                if (_invokeMessageCall(_message, msgHash, gasLimit)) {
-                    _updateMessageStatus(msgHash, Status.DONE);
-                } else {
-                    _updateMessageStatus(msgHash, Status.RETRIABLE);
-                }
-            }
+        if (local.notProcessedByOwner) {
+            uint256 overhead = local.processInTheSameTx //
+                ? GAS_OVERHEAD_RECEIVING_AND_INVOKING
+                : GAS_OVERHEAD_INVOKING;
 
-            // Determine the refund recipient
-            address refundTo =
-                _message.refundTo == address(0) ? _message.destOwner : _message.refundTo;
+            uint256 fee = _calcFee(
+                _message.fee, //
+                _message.gasLimit,
+                local.gas - gasleft() + overhead + _proof.length >> 4,
```

Comment:
will need to require here that the proof is empty for two step withdrawals to make sure that the relayer doesn't charge the user extra for useless data